Dockerfile

```diff
@@ -25,35 +27,33 @@ RUN cp -r build/* /app/static
 RUN mv /app/static/static/* /app/static/
 RUN rm -r /app/static/static
 
+# Set up Django
+FROM python:alpine
 
-FROM python:buster
+# Get PostgreSQL client
+# https://stackoverflow.com/a/47871121
+RUN apk update
+RUN apk add --no-cache postgresql-libs
+RUN apk add --no-cache --virtual build-dependencies gcc musl-dev postgresql-dev
 
-# Copy contents from builder
-COPY --from=builder /app/static /app/static
+# Copy static site from webapp
+COPY --from=webapp /app/static /app/static
 
-# Copy base files
-COPY banwebScrape.py /app/
-COPY course_gather /app/course_gather
-COPY manage.py /app/
-COPY requirements.txt /app/
-COPY scraper /app/scraper
+WORKDIR /app
 
-WORKDIR /app/
+COPY requirements.txt .
 
-# Set environment variables
-ARG DJANGO_SETTINGS_MODULE
-ENV DJANGO_SETTINGS_MODULE=$DJANGO_SETTINGS_MODULE
+# Install Python dependencies
+RUN pip3 install --upgrade pip -r requirements.txt --no-cache-dir
+
+# Remove unneeded build dependencies for smaller image
+RUN apk del build-dependencies
 
-ARG SECRET_KEY
-ENV SECRET_KEY=$SECRET_KEY
+# Set environment variables
+ENV DJANGO_SETTINGS_MODULE="course_gather.settings.prod_settings"
 
-# Install python deps and build backend
-RUN pip3 install --upgrade pip -r requirements.txt
-RUN python3 ./manage.py makemigrations
-RUN python3 ./manage.py migrate
-RUN python3 ./banwebScrape.py
-RUN python3 ./manage.py custom_db_sync /app/writer.csv
+# Copy source
+COPY . .
 
```

Comment:
This file will at minimum need `makemigrations` and `migrate`. SECRET_KEY is also needed for these actions. This will fail out if the database is not already setup (IE the database doesn't have the table, or has the table but not the proper columns)

Without these, the image is in a failing state on first run (until bootstrap).

Reply:
Since the prod Dockerfile uses Postgres (by default), there's not really a good way to init the database while building the image - the image could be built before the database is available, or be built on a different machine entirely.

Reply:
I agree there's no good way. But this is going to cause issues with my deployment cycle as K8s will fail out. Maybe we could do an on run script as a compromise?

Reply:
I don't know if a script that blindly runs `migrate` on every startup is a great idea.

I know very little about K8s but [from a quick search](https://medium.com/@markgituma/kubernetes-local-to-production-with-django-3-postgres-with-migrations-on-minikube-31f2baa8926e) it looks like you might be able to use Jobs to automatically run migrations at certain lifecycle points.

The migrations can be run without actually starting the server, i.e. `docker-compose run backend ./manage.py migrate` or something.

Reply:
If the db is already migrated it will just do nothing. K8s OneTime job could work for that specific use case, but my point here of "this is a failing image on build" still stands.

Reply:
🤷‍♂️ fair enough.

So changing the entrypoint to `startup` where `startup` is a script that
- generates migrations 
- applies migrations
- starts the server 

would work for you?

Reply:
The CMD, yes.
I also see value in using Bash's `set -e` in the startup script (or if shell has an equivalent) which causes the whole script to fail if any components inside it return non-zero.

Reply:
👍 good point.

Check out 020d16e, seems to work fairly well.

Reply:
Very cool! I'm glad sh has that capability as well. I wasn't sure if it would work with just sh.